File: html/app/schemas/attendance.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from datetime import datetime
from app.models.attendance import AttendanceStatus

class AttendanceBase(BaseModel):
    event_id: int = Field(..., example=1)
    student_id: int = Field(..., example=1)
    status: AttendanceStatus = Field(..., example="present")
    mark: Optional[int] = Field(None, ge=1, le=10, example=8, description="Coach's evaluation (1-10)")

class AttendanceCreate(AttendanceBase):
    pass

class AttendanceUpdate(BaseModel):
    status: Optional[AttendanceStatus] = Field(None, example="present")
    mark: Optional[int] = Field(None, ge=1, le=10, example=8)

class AttendanceInDB(AttendanceBase):
    id: int
    model_config = ConfigDict(from_attributes=True)

class AttendanceResponse(AttendanceInDB):
    pass

class BulkAttendanceItem(BaseModel):
    student_id: int = Field(..., example=1)
    status: AttendanceStatus = Field(..., example="present")
    mark: Optional[int] = Field(None, ge=1, le=10, example=8)

class BulkAttendanceCreate(BaseModel):
    event_id: int = Field(..., example=1)
    attendances: List[BulkAttendanceItem]

class StudentAttendanceInfo(BaseModel):
    id: int
    first_name: str
    last_name: str
    model_config = ConfigDict(from_attributes=True)

class EventAttendanceInfo(BaseModel):
    id: int
    start_time: datetime
    type: str
    model_config = ConfigDict(from_attributes=True)

class AttendanceWithDetails(AttendanceResponse):
    student: Optional[StudentAttendanceInfo] = None
    event: Optional[EventAttendanceInfo] = None
    model_config = ConfigDict(from_attributes=True)

class AttendanceStats(BaseModel):
    total_events: int
    present: int
    absent: int
    sick: int
    late: int
    attendance_rate: float

File: html/football_academy_app/lib/models/attendance.dart
class Attendance {
  final int id;
  final int eventId;
  final int studentId;
  final String? studentName;
  final String status;
  final String? notes;

  Attendance({
    required this.id,
    required this.eventId,
    required this.studentId,
    this.studentName,
    required this.status,
    this.notes,
  });

  factory Attendance.fromJson(Map<String, dynamic> json) {
    return Attendance(
      id: json['id'],
      eventId: json['event_id'],
      studentId: json['student_id'],
      studentName: json['student']?['first_name'] != null
          ? '${json['student']['first_name']} ${json['student']['last_name']}'
          : null,
      status: json['status'] ?? 'present',
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'event_id': eventId,
      'student_id': studentId,
      'status': status,
      'notes': notes,
    };
  }
}
